File: ros_wrapper/src/StarMap/tests/test_crop.cpp
#include <sstream>
#include <stdexcept> // std::runtime_error
#include "gtest/gtest.h" // TEST()
#include "starmap/starmap.h" // starmap::

#include "opencv2/opencv.hpp" // cv::*
#include "opencv2/imgcodecs.hpp" // cv::imread, cv::imwrite

class FileNotFoundError : std::runtime_error {
public:
  FileNotFoundError(const std::string& what_arg ) : std::runtime_error( what_arg ) { }
  FileNotFoundError(const char* what_arg ) : std::runtime_error( what_arg ) { }
};

cv::Mat imread_img(const cv::Size img_shape,
                   const std::string fname = "tests/data/lena512.pgm") {
  cv::Mat outimg;
  cv::Mat inimg = cv::imread(fname, cv::IMREAD_UNCHANGED);
  if (inimg.data == nullptr)
    throw new FileNotFoundError(fname);
  cv::resize(inimg, outimg, img_shape);
  return outimg;
}


/**
 * @param img_shape: (height, width)
 */
bool _test_crop_lenna(const int rows,
                      const int cols,
                      const int desired_side = 256,
                      const int rot = 0) {
  cv::Size imgshape(rows, cols);
  auto img = imread_img(imgshape);
  auto cropped = starmap::crop(img, desired_side);
  std::ostringstream filepath;
  filepath << "tests/data/test-crop-lenna-" << rows << "-" << cols << ".pgm";
  // cv::imwrite(filepath.str(), cropped);
  cv::Mat expected_img = cv::imread(filepath.str(), cv::IMREAD_UNCHANGED);
  cv::Mat diff = cropped != expected_img;
  return cv::countNonZero(diff) == 0;
}


TEST(CropTest, HandlesFatBigBig) {
  ASSERT_TRUE(_test_crop_lenna(301, 401)) << "Images do not match";
}


TEST(CropTest, HandlesFatSmallBig) {
  ASSERT_TRUE(_test_crop_lenna(101, 401)) << "Images do not match";
}

TEST(CropTest, HandlesFatSmallSmall) {
  ASSERT_TRUE(_test_crop_lenna(101, 201)) << "Images do not match";
}

TEST(CropTest, HandlesTallSmallSmall) {
  ASSERT_TRUE(_test_crop_lenna(201, 101)) << "Images do not match";
}

TEST(CropTest, HandlesTallBigSmall) {
  ASSERT_TRUE(_test_crop_lenna(401, 101)) << "Images do not match";
}

TEST(CropTest, HandlesTallBigBig) {
  ASSERT_TRUE(_test_crop_lenna(401, 301)) << "Images do not match";
}

TEST(CropTest, CarCropTest) {
  constexpr bool DEBUG = false;
  const std::string infname("tests/data/car.jpg");
  const std::string expectedfname("tests/data/car-cropped.jpg");
  const int desired_side = 256;
  cv::Mat inimg = cv::imread(infname, cv::IMREAD_COLOR);
  if (inimg.data == nullptr)
    throw new FileNotFoundError(infname);
  auto cropped = starmap::crop(inimg, desired_side);
  cv::Mat expected_img = cv::imread(expectedfname, cv::IMREAD_COLOR);
  if (DEBUG) {
    cv::imshow("cropped", cropped);
    cv::imshow("expected_img", expected_img);
  }
  cv::Mat cropped_gray, expected_img_gray;
  cv::cvtColor(cropped,  cropped_gray, cv::COLOR_BGR2GRAY);
  cv::cvtColor(expected_img, expected_img_gray, cv::COLOR_BGR2GRAY);
  cv::Mat diff;
  cv::absdiff(cropped_gray, expected_img_gray, diff);

  if (DEBUG) {
    cv::imshow("diff", diff);
    cv::waitKey(-1);
  }
  double diffval = cv::sum(diff)[0];
  ASSERT_TRUE( diffval < 0.02 * 255 * desired_side * desired_side) <<
    "diffval: " << diffval << " < exp: " << 0.02 * 255 * desired_side * desired_side;
}

File: ros_wrapper/src/StarMap/tests/test_crop.py
from pathlib import Path
import logging
logging.basicConfig(level=logging.DEBUG,
                    format='%(asctime)-15s %(funcName)-15s %(message)s')
LOG = logging.getLogger(__name__)

import cv2
import numpy as np

from starmap.utils.img import Crop, NewCrop

DEBUG = False

def random_img(img_shape):
    img = np.random.randint(255, size=(img_shape[0], img_shape[1], 3), dtype=np.uint8)
    return img

def imread_img(img_shape,
               fname=Path(__file__).parent / 'data' / 'lena512.pgm'):
    return cv2.resize(cv2.imread(str(fname)), img_shape)


def _test_crop(img_shape, desired_side=256, rot=0, img_gen=imread_img,
               new_crop_kw={}, skip_match=False):
    """
    @param img_shape: (height, width)
    """
    img = img_gen(img_shape)
    center = np.array([img.shape[1] // 2, img.shape[0] // 2])
    max_side = max(img.shape[:2])
    new_img = Crop(img, center, max_side, 0, desired_side)
    proposed_img = NewCrop(img, center, max_side, 0, desired_side, **new_crop_kw)
    if DEBUG and not (new_img != proposed_img).sum() / np.prod(new_img.shape) < 0.02:
        cv2.imshow("Crop", new_img)
        cv2.imshow("NewCrop", proposed_img)
        cv2.imshow("diff", new_img - proposed_img)
        cv2.waitKey(-1)
        LOG.warning("Images did not match with %d pixels" % np.sum(new_img != proposed_img))
        #assert False
    elif not skip_match:
        assert (new_img != proposed_img).sum() / np.prod(new_img.shape) < 0.02


def test_crop_lena():
    _test_crop((300, 400), img_gen=imread_img)


def test_crop_fat_big_big_no_rot():
    _test_crop((300, 400))


def test_crop_fat_small_big_no_rot():
    _test_crop((100, 400))


def test_crop_fat_small_small_no_rot():
    _test_crop((100, 200))


def test_crop_tall_small_small_no_rot(new_crop_kw=dict()):
    _test_crop((200, 100), new_crop_kw=new_crop_kw)


def test_crop_tall_big_small_no_rot():
    _test_crop((400, 100))


def test_crop_tall_big_big_no_rot():
    _test_crop((400, 300))


def test_crop_tall_big_small_no_rot():
    _test_crop((800, 100))


def test_crop_tall_big_big_no_rot():
    _test_crop((800, 750))

def test_crop_odd_size():
    _test_crop((213, 304), skip_match=True)


if __name__  == '__main__':
    ## Succeeds
    test_crop_lena()
    test_crop_fat_big_big_no_rot()
    test_crop_fat_small_big_no_rot()
    test_crop_fat_small_small_no_rot()
    test_crop_tall_small_small_no_rot()
    test_crop_tall_big_small_no_rot()
    test_crop_tall_big_big_no_rot()
    test_crop_tall_big_small_no_rot()
    test_crop_tall_big_big_no_rot()
    test_crop_odd_size()
